plugins/catalog-backend/src/service/CatalogBuilder.ts

```diff
@@ -213,138 +188,60 @@ export class CatalogBuilder {
-   * but before being validated and post-processed.
+   * Adds, or overwrites, a handler for placeholders (e.g. $file) in entity
+   * definition files.
    *
-   * @param processors One or more processors
+   * @param key The key that identifies the placeholder, e.g. "file"
+   * @param resolver The resolver that gets values for this placeholder
    */
-  addPreProcessor(...processors: CatalogProcessor[]): CatalogBuilder {
-    this.preProcessors.push(...processors);
+  setPlaceholderResolver(
+    key: string,
+    resolver: PlaceholderResolver,
+  ): CatalogBuilder {
+    this.placeholderResolvers[key] = resolver;
     return this;
   }
 
   /**
-   * Sets what processors to run after each entity has been read and parsed,
-   * but before being validated and post-processed.
+   * Sets the validator function to use for one or more special fields of an
+   * entity. This is useful if the default rules for formatting of fields are
+   * not sufficient.
    *
-   * This function replaces the default set of processors in this stage; use
-   * with care.
+   * This function has no effect if used together with
+   * {@link CatalogBuilder#replaceEntityPolicies}.
    *
-   * @param processors One or more processors
+   * @param validators The (subset of) validators to set
    */
-  replacePreProcessors(processors: CatalogProcessor[]): CatalogBuilder {
-    this.preProcessors = [...processors];
-    this.preProcessorsReplace = true;
+  setFieldFormatValidators(validators: Partial<Validators>): CatalogBuilder {
+    lodash.merge(this.fieldFormatValidators, validators);
     return this;
   }
 
   /**
-   * Adds processors that run after each entity has been read, parsed,
-   * run through the pre-processors, and validated.
+   * Adds entity processors. These are responsible for reading, parsing, and
+   * processing entities before they are persisted in the catalog.
    *
    * @param processors One or more processors
    */
-  addPostProcessor(...processors: CatalogProcessor[]): CatalogBuilder {
-    this.postProcessors.push(...processors);
+  addProcessor(...processors: CatalogProcessor[]): CatalogBuilder {
+    this.processors.push(...processors);
     return this;
   }
 
   /**
-   * Sets what processors to run after each entity has been read, parsed,
-   * run through the pre-processors, and validated.
+   * Sets what entity processors to use. These are responsible for reading,
+   * parsing, and processing entities before they are persisted in the catalog.
    *
-   * This function replaces the default set of processors in this stage; use
-   * with care.
+   * This function replaces the default set of processors; use with care.
    *
    * @param processors One or more processors
    */
-  replacePostProcessors(processors: CatalogProcessor[]): CatalogBuilder {
-    this.postProcessors = [...processors];
-    this.postProcessorsReplace = true;
+  replaceProcessors(processors: CatalogProcessor[]): CatalogBuilder {
```

Comment:
still just a big replace - no fancy replace-single-things method yet, design of that is tbd